src/Components/test/testassets/TestServer/Program.cs

```diff
@@ -1,19 +1,90 @@
+using Microsoft.AspNetCore.Http.Features;
 using Microsoft.Extensions.Configuration;
 using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
 using Microsoft.Extensions.Logging;
 using Microsoft.Extensions.Logging.Testing;
+using DevServerProgram = Microsoft.AspNetCore.Blazor.DevServer.Server.Program;
 
 namespace TestServer
 {
     public class Program
     {
         public static void Main(string[] args)
         {
-            BuildWebHost(args).Run();
+            var createIndividualHosts = new Dictionary<string, (IWebHost host, string basePath)>
+            {
+                ["Client authentication"] = (BuildWebHost<AuthenticationStartup>(CreateAdditionalArgs(args)), "/subdir"),
+                ["Server authentication"] = (BuildWebHost<ServerAuthenticationStartup>(CreateAdditionalArgs(args)), "/subdir"),
+                ["CORS (WASM)"] = (BuildWebHost<CorsStartup>(CreateAdditionalArgs(args)), "/subdir"),
+                ["Prerendering (Server-side)"] = (BuildWebHost<PrerenderedStartup>(CreateAdditionalArgs(args)), "/prerendered"),
+                ["Globalization + Localization (Server-side)"] = (BuildWebHost<InternationalizationStartup>(CreateAdditionalArgs(args)), "/subdir"),
+                ["Server-side blazor"] = (BuildWebHost<ServerStartup>(CreateAdditionalArgs(args)), "/subdir"),
+                ["Hosted client-side blazor"] = (BuildWebHost<ClientStartup>(CreateAdditionalArgs(args)), "/subdir"),
+                ["Dev server client-side blazor"] = CreateDevServerHost(CreateAdditionalArgs(args))
+            };
+
+            var mainHost = BuildWebHost(args);
+
+            Task.WhenAll(createIndividualHosts.Select(s => s.Value.host.StartAsync())).GetAwaiter().GetResult();
+
+            var testAppInfo = mainHost.Services.GetRequiredService<TestAppInfo>();
+            testAppInfo.Scenarios = createIndividualHosts
+                .ToDictionary(kvp => kvp.Key,
+                kvp => kvp.Value.host.ServerFeatures.Get<IServerAddressesFeature>().Addresses.FirstOrDefault()
+                    .Replace("127.0.0.1", "localhost") + kvp.Value.basePath);
+
+            mainHost.Run();
+        }
+
+        private static (IWebHost host, string basePath) CreateDevServerHost(string[] args)
+        {
+            var contentRoot = typeof(Program).Assembly.GetCustomAttributes<AssemblyMetadataAttribute>()
+                .Single(a => a.Key == "Microsoft.AspNetCore.Testing.BasicTestApp.ContentRoot")
+                .Value;
+
+            var finalArgs = args.Concat(new[]
+            {
+                "--contentroot", contentRoot,
+                "--pathbase", "/subdir",
+                "--applicationpath", typeof(BasicTestApp.Program).Assembly.Location,
+            }).ToArray();
+            var host = DevServerProgram.BuildWebHost(finalArgs);
+            return (new WebHostShim(host), "/subdir");
         }
 
+        private class WebHostShim : IWebHost
```

Comment:
This is temporary, I'm planning to send out a separate PR where I convert all the test infrastructure to use IHost. I just didn't want to increase the scope of this PR even more.